write.go

```diff
@@ -232,12 +253,184 @@ func writeProduceRequestV2(w *bufio.Writer, codec CompressionCodec, correlationI
 	writeArrayLen(w, 1)
 	writeInt32(w, partition)
 	writeInt32(w, size)
+	w.Write(msgBuf)
+	return w.Flush()
+}
+
+func writeMessageSet(codec CompressionCodec, correlationID int32, clientId, topic string, partition int32, timeout time.Duration, requiredAcks int16, msgs ...Message) ([]byte, error) {
+	var size int32
+	attributes := int8(CompressionNoneCode)
+
+	// if compressing, replace the slice of messages with a single compressed
+	// message set.
+	if codec != nil {
+		var err error
+		if msgs, err = compress(codec, msgs...); err != nil {
+			return nil, err
+		}
+		attributes = codec.Code()
+	}
 
 	for _, msg := range msgs {
-		writeMessage(w, msg.Offset, attributes, msg.Time, msg.Key, msg.Value)
+		size += 8 + // offset
+			4 + // message size
+			4 + // crc
+			1 + // magic byte
+			1 + // attributes
+			8 + // timestamp
+			sizeofBytes(msg.Key) +
+			sizeofBytes(msg.Value)
 	}
 
-	return w.Flush()
+	buf := &bytes.Buffer{}
+	buf.Grow(int(size))
+	bufWriter := bufio.NewWriter(buf)
```

Comment:
This is likely going to introduce a significant performance regression on producers which use kafka-go. This code is introducing two new temporary buffers on code which we have on hot paths of our data pipelines (I'd assume other companies do as well). Those buffers will result in two extra copies of the message set, and increase the memory footprint of kafka producers. We need to introduce this change without taking the risk to suffer performance regressions.

Reply:
Ok. I'll try to do something about it.